### PR TITLE
Correctly align messages in right to left languages

### PR DESCRIPTION
**Pull request by aaronraimist**
Fixes https://github.com/vector-im/element-web/issues/4771

screenshots

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Correctly align messages in right to left languages ([\#5453](https://github.com/matrix-org/matrix-react-sdk/pull/5453)). Fixes vector-im/element-web#4771. Contributed by @aaronraimist.

### CODE REVIEW

File: res/css/views/rooms/_EventTile.scss
Comment (jryans) [inline res/css/views/rooms/_EventTile.scss:298]: Since this comment is only about the `overflow-y` part, please move it either above that line or inline to the right.

File: res/css/views/rooms/_EventTile.scss
Comment (jryans) [inline res/css/views/rooms/_EventTile.scss:301]: Rather than changing `mx_EventTile_body` elements (which are `span`s) to `display: block`, it would be more straightforward to instead make them `div`s and get this for free.

It would also make sense to me to change `mx_EventTile_content` elements to `div`s as well, and remove their `display: block` style, so we are consistently using `div` vs. `span` for their intended meaning.

**Review by jryans — CHANGES_REQUESTED** [review]
Thanks for working on this! 😄 Overall, it seems like a good improvement, but I have a few nits on the details.

File: res/css/views/rooms/_EventTile.scss
Comment (aaronraimist) [reply res/css/views/rooms/_EventTile.scss]: Okay I've changed them over to `div`s. This mostly works except it breaks the layout for edited messages and `/me` messages. I'm not sure how to fix this.

screenshots

Comment (jryans) [reply res/css/views/rooms/_EventTile.scss]: Ah, hmm... 😖 Okay, I'll have to think about how we'd want to handle the edited case.

**Comment by ahangarha** [comment]
It is needed to add `dir="auto"` HTML attribute to text placeholders. As far as I know this is not something to be solved effetively using only css.

**Comment by aaronraimist** [comment]
Yes that already exists. The problem is it doesn’t do anything on a span element.

**Comment by ahangarha** [comment]
> Yes that already exists. The problem is it doesn’t do anything on a span element.

I saw it just now on element desktop. I just modified the css with dev tools and the initial result is fine:

``` css
.mx_EventTile_body {
    overflow-y: hidden; /* already exists */
    text-align: start;
    display: block;
}
```

image
I indicated rightly/wrongly aligned lines with some color coding on side. Those two red lines are wrong because I made those message using shift+enter so that they are posed in a single span. This is fine to me as it is manageable and understandable by the user. To me this is fine. I will try to check more complex cases.

**Comment by ahangarha** [comment]
I have noticed issue with lists which can be resolved using this css:

```
.markdown-body ol, .markdown-body ul {
    padding-top: 0;
    padding-bottom: 0;
    padding-block-start: 2em;
}
```
result:

image

**Comment by ahangarha** [comment]
Quotation fix:
```css
.markdown-body blockquote {
    margin: 0 0 16px;
    padding: 0 15px;
    color: #777;
    /* border-left: 4px solid #ddd; */
    border-inline-start: 4px solid #ddd;
}
```

before:

image

after:

image

File: res/css/views/rooms/_EventTile.scss
Comment (ahangarha) [reply res/css/views/rooms/_EventTile.scss]: what is the issue with edited case?

Comment (aaronraimist) [reply res/css/views/rooms/_EventTile.scss]: See the screenshot above. The `(edited)` label is in the wrong place.

**Comment by jryans** [comment]
This is still in my mental queue to think about, but it take a bit of time / some conversations with Design to work out exactly what we want here.

File: res/css/views/rooms/_EventTile.scss
Comment (ahangarha) [reply res/css/views/rooms/_EventTile.scss]: why not set its display to block and then text-align to start?

**Comment by ahangarha** [comment]
Let me know if I can help.

**Comment by ahangarha** [comment]
What is the update? What is preventing this PR to be finalized and merged?

**Comment by jryans** [comment]
I still need to work out an approach that handles edited markers and such, and then review that with Design. I haven't had time to do that yet.

**Comment by ahangarha** [comment]
> I still need to work out an approach that handles edited markers and such, and then review that with Design. I haven't had time to do that yet.

May you share what specific issue is there? One (maybe me) might help

**Comment by jryans** [comment]
I have unfortunately failed to find the proper time to think about this... 😓 I'll send this back to the general queue in hopes that it might lead to new input here.

**Comment by ahangarha** [comment]
> I have unfortunately failed to find the proper time to think about this... sweat I'll send this back to the general queue in hopes that it might lead to new input here.

May you please share what is there to think? Share the issue so that others might be able to offer some help.

**Comment by t3chguy** [comment]
@ahangarha its already in a thread above.
https://github.com/matrix-org/matrix-react-sdk/pull/5453#discussion_r533035486

**Comment by turt2live** [comment]
I'm obviously not familiar with RTL languages, but would it be awful to keep the `(edited)` bit on the right side of the message text? If that's fine, then we might be able to throw in some flexbox or similar CSS trickery to keep everything lined up (a problem would be multiline messages, but that might be solved with less flexbox and more DOM nesting)

**Comment by ahangarha** [comment]
> @ahangarha its already in a thread above.
> [#5453 (comment)](https://github.com/matrix-org/matrix-react-sdk/pull/5453#discussion_r533035486)

If this is the only remaining issue, I will share a solution. Just please list all the remaining issue.

**Comment by aaronraimist** [comment]
@turt2live My understanding is ideally it would be on the left but if we could at least get the message oriented correctly and have the (edited) in the wrong spot that would still be much better than how it currently works.

**Comment by jryans** [comment]
At the moment, the app currently (without the change here) shows the "(edited)" label inline with the message text. The "easy" fix to apply RTL to message also causes it become a block element, which moves the "(edited)" label. I believe the main options to move forward are either:

* Find a technical solution which preserve the inline placement of the edited label
* Work with Design team to change the placement to edited label

**Comment by ahangarha** [comment]
As far as I see, it can be tweaked to follow bidi algorithm. So, if it is Latin, it remains on left and if gets translated to Persian (or other RTL languages) it moves to right. I think this is the right behavior. If you want to force it to follow the direction of the related text, it is doable but there should be significant change in the layout.

Yet to be frank, this is a minor issue as compared to the situation we (RTL people) face with. I would prefer to get the current improved version and then in next version improve this little minor issue.

Is there any build based on the current modifications so that I can check different scenarios?

**Comment by jryans** [comment]
> Is there any build based on the current modifications so that I can check different scenarios?

See the [deploy preview](https://deploy-preview-5453--matrix-react-sdk.netlify.app/), you can find the latest link in the Checks section of the PR.

**Comment by amshakal** [comment]
Hey! Just checked the deloyed preview, I think it might be better to align (edited) to the right side for edited RTL scripts. Usually, for designing in RTL it's preferred to mirror all components but I'm guessing that doesn't make any sense if we are using multi languages in the timeline. 

Did you need feedback on anything else specifically?

**Comment by ahangarha** [comment]
> See the [deploy preview](https://deploy-preview-5453--matrix-react-sdk.netlify.app/), you can find the latest link in the Checks section of the PR.

I check the preview. It looks fine in most cases. Still I should check more. But I found out one issue with code section.

image

It should not follow the the surrounding text direction. It should independently detect the direction of text in it.

**Comment by ahangarha** [comment]
Guys,

I have made these simple styles which nearly all issues with bidi support: https://userstyles.world/style/2159/element-bidi

As far as I see it is working perfectly. It was little issue if we have a mixed markdown content which can be managed in some way but doesn't have periority now.

I wonder why still we don't have this feature on element!

image

**Comment by aaronraimist** [comment]
@ahangarha it looks like it has basically the same problem with the `(edited)` label though

**Comment by ahangarha** [comment]
> @ahangarha it looks like it has basically the same problem with the `(edited)` label though

I really don't understand why this is that much important to keep the 99.9% fix on hold.

See. I am using Element and I am who uses RTL/LTR both. If anyone is going to suffer it would be me and people like me. And I am fine with this. I don't see it even as a problem.

This is not a security issue! Just keep this to be fixed (if needed at all) in future. Don't hold this major improvement just for a tiny issue which is not even noticeable!

**Comment by aaronraimist** [comment]
> See. I am using Element and I am who uses RTL/LTR both. If anyone is going to suffer it would be me and people like me. And I am fine with this. I don't see it even as a problem.

But that's not true. If it only broke on RTL text it would probably be fine. But your CSS (as well as this PR) move the edited label to the wrong place on all messages, not just RTL.

**Comment by ahangarha** [comment]
> > See. I am using Element and I am who uses RTL/LTR both. If anyone is going to suffer it would be me and people like me. And I am fine with this. I don't see it even as a problem.
> 
> But that's not true. If it only broke on RTL text it would probably be fine. But your CSS (as well as this PR) move the edited label to the wrong place on all messages, not just RTL.

How much wrong is that place? It is just going to the next line. Is it a big issue? bigger than messing up with mixed RTL/LTR text? I don't think so. And frankly I think we should move that entirely out of that place but that is a different issue.

The fact is that Element has a big issue when it comes to RTL text. We can fix this big issue which would produce a little visual issue for only edited message. I think logic says lets fix the bigger problem now and fix the smaller later. We shouldn't let the smaller issue prevent us from fixing the bigger issue. Am I wrong?

**Comment by danialbehzadi** [comment]
I completly agree with @ahangarha in this case. Making Element usable in BiDi is more important than a minor visual issue. Anyway I think the current edited sign behaviour is already a bad UX decision and we should find another approach for indicating a message is edited.

**Comment by SeedPuller** [comment]
As a RTL user, if this "edited" sign thing is the only reason that prevents me to use my language in messages properly, I really don't care about it.So I totally agree with @ahangarha  and @danialbehzadi .

**Comment by luixxiul** [comment]
@ahangarha @danialbehzadi @SeedPuller Since a couple of days I was trying to edit styles to make UI compatible with RTL languages. Please check https://github.com/vector-im/element-web/issues/14520#issuecomment-1086704259 for a short video of UI in Hebrew (as UI is not available in languages of Arabic alphabet yet).

**Comment by ahangarha** [comment]
Thank you for your work but the issue is not to make overall layout RTL. We need it for sure but we need to show messages in correct direction regardless of the global language and direction.

در ۳ آوریل ۲۰۲۲ ۹:۵۴:۵۶ (GMT+03:00)، Suguru Hirahara ***@***.***> نوشت:
***@***.*** @danialbehzadi @SeedPuller Since a couple of days I was trying to edit styles to make UI compatible with RTL languages. Please check https://github.com/vector-im/element-web/issues/14520#issuecomment-1086704259 for a short video of UI in Hebrew (as UI is not available in languages of Arabic alphabet yet).

**Comment by luixxiul** [comment]
@ahangarha From my understanding `(edited)` label breaking can be effectively avoided by setting `dir="auto"` to the div and changing the margin setting of the `(edited)` span wrapped with the div. Since I don't want to override this PR, I am going to submit the change with another PR to address https://github.com/vector-im/element-web/issues/14520 partially.

after

Mind the place of the two `(edited)` labels.

**Comment by ahangarha** [comment]
By breaking I think they mean it is not inline anymore. To me it is not a problem at all and even if it is, it is not something to block such an important change.

**Comment by luixxiul** [comment]
> By breaking I think they mean it is not inline anymore. 

It is going to be taken care of :wink: 
after

**Comment by alongls** [comment]
Hello
we are anticipating this improvements and finally to see the Hebrew and Arabic messages in the correct side
thanks.

**Comment by andybalaam** [comment]
This came up in our weekly review as the oldest open non-draft PR! If someone wanted to help it get over the line there would be kudos!

**Comment by andybalaam** [comment]
This PR adds potentially really valuable functionality to Element Web, but it is currently blocked because it breaks layout in non-RTL contexts (see [this comment](https://github.com/matrix-org/matrix-react-sdk/pull/5453#issuecomment-983498831).

It also need updating to resolve merge conflicts with the latest code.

Is anyone willing to step up to fix the remaining problems so we can merge this PR? If not, I will convert this to draft to indicate that it is not ready, and not under active development.

Lots of people will thank you if you give this a try!

**Comment by andybalaam** [comment]
Marking this as draft because it needs changes and no-one is actively working on it. If you want to work on it, please do, and mark as Ready for Review when you have address the changes needed. Feel free to discuss what is needed here or in #element-dev:matrix.org Matrix room.

**Comment by andybalaam** [comment]
Hi @ahangarha , thank you for your patience with this!

ahangarha:
> There are many questions about the importance of the break and almost there is no answer to it.

My understanding of the situation is that this change introduces a bug.

Quoting aaronraimist:
> [this PR] moves the edited label to the wrong place on all messages, not just RTL.

As for how important it is: the answer is "important enough to need fixing before we can merge this change".

> I would suggest to set a meeting (or some other effective way of communication) to discuss this issue.

I'm not sure whether that is needed here: if someone fixes the bug (and brings this change up-to-date, which I'm afraid will probably be harder) then we will review it and expect to merge it.

**Comment by ahangarha** [comment]
Sounds good. Though I think it is more than my current knowledge, but I try to see if I can fix the conflict. I might be able to do it on this weekend. Let's see. At least I try :)

**Comment by andybalaam** [comment]
Awesome, good luck!

**Comment by luixxiul** [comment]
I made a quick implementation at https://github.com/matrix-org/matrix-react-sdk/pull/9026. Feedback for https://pr9026--matrix-react-sdk.netlify.app/ is appreciated.

**Comment by dbkr** [comment]
I've had a go at a version of this PR that keeps the 'edited' marker in the correct place at https://github.com/matrix-org/matrix-react-sdk/pull/12837

**Comment by dbkr** [comment]
https://github.com/matrix-org/matrix-react-sdk/pull/12837 is merged which should fix this! 🤞

**Pull request closed without merging**
